### ISSUE TITLE
Breaking change in SSP 1.14.7

### ISSUE THREAD

**Issue by moffe42** [simplesamlphp/simplesamlphp#616]
I recently tried to update from 1.14.6 to 1.14.14, but was unable to. I got the following error:

```
Fatal error:  Uncaught TypeError: Argument 1 passed to SimpleSAML_Session::load() must be an instance of SimpleSAML_Session, boolean given, called in /mnt/www/jacob-login/lib/SimpleSAML/Session.php on line 266 and defined in /mnt/www/jacob-login/lib/SimpleSAML/Session.php:379
Stack trace:
#0 /mnt/www/jacob-login/lib/SimpleSAML/Session.php(266): SimpleSAML_Session::load(false)
#1 /mnt/www/jacob-login/lib/SimpleSAML/Logger.php(250): SimpleSAML_Session::getSessionFromRequest()
#2 [internal function]: SimpleSAML_Logger::flush()
#3 {main}
  thrown in /mnt/www/jacob-login/lib/SimpleSAML/Session.php on line 379
```

This is due to commit 977b8e869898b32052e70c4f0d1c58c6369222c7 that was done 10 days after the release of 1.14.6. The commit changed the was sessions was serialized. So sessions that was serialized under 1.14.6 cannot be unserialized in 1.14.7 and above.

Or am I missing something??

**Comment by mcouillard** [comment]
I ran into this, too. I got around it during my upgrade from 1.14.1 to 1.14.14 by changing the session IDs so old sessions are not re-used.

**Comment by moffe42** [comment]
The problem is that this do not fix existing sessions. So the only solution is to invalidate all existing sessions, which is not a good solution.

**Comment by jaimeperez** [comment]
Hi @moffe42!

As far as I recall, sessions failed to load when they were created with >=1.14.7 but loaded with <1.14.7, but the other way around just an error was logged but the sessions were _converted_ automatically. The key here is: if you have more than one instance of SSP accessing the same session backend, update all of them at the same time.

Other than that (I insist, this is to the best of my memory), this was not a BC break.

**Comment by moffe42** [comment]
The facts is updating from 1.14.6 do not work. My users are meet with a SSP error page, when we update beyond 1.14.7. This constitutes as a BC breakage in my world. Especially when only the patch version was changed.

**Comment by jaimeperez** [comment]
Hi @moffe42!

I understand you are using the PHP session handler, right? There's a bug there that I just fixed, and that's why you were experiencing a different behaviour than the one we experienced in Feide, and the one I saw in my tests before releasing 1.14.7. It would have been useful to have more information  about your setup and the errors your users were seeing in order to diagnose the problem, but after testing it for a while myself I noticed the problem anyway.

In any case, the old sessions will be invalidated. There's simply no way around that, unfortunately. There was no way to solve #424 without breaking horribly the API or invalidating sessions, so I took the least harmful way.

@thijskh, do you mind releasing a new 1.14.15 version? I can't do it myself right now, and won't be able for a while. You just need to follow the same steps as usual (updating the changelog, upgrade notes if necessary, changing the version number in `SimpleSAML_Configuration` and the RPM spec file, signing the commit and tag and creating the release here). SSP's website will update automatically (except for the front page, which I can update in a couple minutes when you are done), and then you just need to send an email to the SSP announce and users mailing lists. Thanks a lot in advance!

**Comment by moffe42** [comment]
@jaimeperez we are using https://github.com/ColourboxDevelopment/simplesamlphp-module-redis as store and out users are getting the normal SSP error page with an Id. And as far as I can gather from the already implemented stores and and the abstract class SimpleSAML_Store. I might have missed something, but according the the docs (inline code comments), it should be behaving correctly. (I used SimpleSAML_Memcached as reference for behaviour).

**Comment by jaimeperez** [comment]
Hi @moffe42!

I see there's the [same issue](https://github.com/ColourboxDevelopment/simplesamlphp-module-redis/blob/master/lib/Store/Redis.php#L52) in that store as in the PHP store. Basically, you just return whatever `unserialize()` returns, while `SimpleSAML_Session` was using `null` as a sign that there was no session. Unfortunately, `unserialize()` returns `false` if something goes wrong, which is not `null`, and therefore you were bitten by the issue as well.

In any case, the changes I applied should fix it. I'll try to have it in a bugfix release ASAP.

**Comment by jaimeperez** [comment]
This was fixed in the 1.14.15 stable release.
